src/firmware/app/control/trajectory_planner.c

```diff
@@ -228,155 +128,379 @@ app_trajectory_planner_interpolateConstantPeriodPositionTrajectory(
+    }
 
-    for (unsigned int i = 0; i < num_segments; i++)
+    // Calculate the new final velocity based on the initial velocity, segment length, and
+    // the desired segment duration
+    const float next_speed =
+        (2 * segment_length / desired_segment_duration) + current_speed;
+
+    if (rebalance_angular)
+    {
+        trajectory->trajectory_elements[trajectory_index].angular_speed = next_speed;
+    }
+    else
     {
-        // Reverse the positions
-        reverse[(num_segments - 1) - i].position =
-            forwards_trajectory->trajectory_elements[i].position;
-        reverse[(num_segments - 1) - i].time =
-            path_duration - forwards_trajectory->trajectory_elements[i].time;
-
-        // Reverse the speed profile
-        reverse_speed_profile[(num_segments - 1) - i] =
-            forwards_trajectory->speed_profile[i];
+        trajectory->trajectory_elements[trajectory_index].linear_speed = next_speed;
     }
-    // Copy reverse element array back
-    for (unsigned int i = 0; i < num_segments; i++)
+}
+
+void app_trajectory_planner_getMaxAllowableSpeedProfile(
+    Polynomial2dOrder3_t path, const unsigned int num_elements, const float t_start,
+    const float t_end, const float max_allowable_acceleration,
+    float* max_allowable_speed_profile)
+{
+    Polynomial2dOrder2_t first_deriv = shared_polynomial2d_differentiateOrder3(path);
+    Polynomial2dOrder1_t second_deriv =
+        shared_polynomial2d_differentiateOrder2(first_deriv);
+    const float delta_t = (t_end - t_start) / num_elements;
+
+    for (unsigned int i = 0; i < num_elements; i++)
     {
-        forwards_trajectory->trajectory_elements[i].position = reverse[i].position;
-        forwards_trajectory->trajectory_elements[i].time     = reverse[i].time;
-        forwards_trajectory->speed_profile[i]                = reverse_speed_profile[i];
+        float t = t_start + i * delta_t;
+
+        // Create the polynomial representing path curvature
+        //                                              1
+        //                              ---------------------------------
+        //                                     abs(x'y'' - y'x'')
+        //        radius of curvature =      ----------------------
+        //                                     (x'^2 + y'^2)^(3/2)
+        //
+        const float numerator =
+            fabs(shared_polynomial1d_getValueOrder2(first_deriv.x, t) *
+                     shared_polynomial1d_getValueOrder1(second_deriv.y, t) -
+                 shared_polynomial1d_getValueOrder2(first_deriv.y, t) *
+                     shared_polynomial1d_getValueOrder1(second_deriv.x, t));
+        const float denominator =
+            pow(pow(shared_polynomial1d_getValueOrder2(first_deriv.x, t), 2) +
+                    pow(shared_polynomial1d_getValueOrder2(first_deriv.y, t), 2),
+                3.0 / 2.0);
+
+        const float radius_of_curvature = 1 / (numerator / denominator);
```

Comment:
Can we this be moved to firmware/shared/math?

Reply:
I can - but I highly doubt the math will be used elsewhere in the firmware. What are peoples thoughts on this?

@MathewMacDougall @akhilveeraghanta @garethellis0

Reply:
This function didn't actually change in this PR, so is it cool if I just file a ticket @jonathanlew ?